<a href="https://colab.research.google.com/github/Jorge2018/DataScience2022-2/blob/main/Script_knnAbalon.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [66]:
import pandas as pd
import numpy as np
# For scaling data
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.datasets import load_iris
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score, precision_score, recall_score, f1_score


df=pd.read_csv(r'https://raw.githubusercontent.com/Jorge2018/RepositorioArchivos/main/abalone.data')
lista=['Sex','Length','Diameter','Height','Whole weight','Shucked weight','Viscera weight','Shell weight','Rings']
df.columns=lista
df.head()

Sex  Length  Diameter  Height  Whole weight  Shucked weight  Viscera weight  \
0   M   0.350     0.265   0.090        0.2255          0.0995          0.0485   
1   F   0.530     0.420   0.135        0.6770          0.2565          0.1415   
2   M   0.440     0.365   0.125        0.5160          0.2155          0.1140   
3   I   0.330     0.255   0.080        0.2050          0.0895          0.0395   
4   I   0.425     0.300   0.095        0.3515          0.1410          0.0775   

   Shell weight  Rings  
0         0.070      7  
1         0.210      9  
2         0.155     10  
3         0.055      7  
4         0.120      8

**Encuentren los nombres de clase y determinen cómo están las clases balanceadas**

In [67]:
df['Sex'].value_counts(normalize = True)

M    0.365661
I    0.321360
F    0.312979
Name: Sex, dtype: float64

**Las clases se ven bien balanceadas existiendo en promedio 3 pts de diferencia entre clases**

**Conviertan los nombres de cadenas de las clases para valores numéricos.**

In [50]:
df['Sex'] = df['Sex'].replace({'M': 0, 'I': 1,'F':2})
df['Sex']

0       0
1       2
2       0
3       1
4       1
       ..
4171    2
4172    0
4173    0
4174    2
4175    0
Name: Sex, Length: 4176, dtype: int64

**Separa X, y**

In [51]:
y = df['Sex']
X = df.drop(columns = 'Sex')
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [52]:
#Instancia del Scaler
scaler = StandardScaler()


In [53]:
# Instancia del modelo KNN (el parametro n_neighbors es modificable)
knn = KNeighborsClassifier(n_neighbors=30)
# Revision de hiperparametros
knn.get_params()

{'algorithm': 'auto',
 'leaf_size': 30,
 'metric': 'minkowski',
 'metric_params': None,
 'n_jobs': None,
 'n_neighbors': 30,
 'p': 2,
 'weights': 'uniform'}

In [54]:
#instancia y entrenamiento del pipeline que contiene el scaler y knn
knn_pipe = make_pipeline(scaler, knn)
knn_pipe.fit(X_train, y_train)


Pipeline(steps=[('standardscaler', StandardScaler()),
                ('kneighborsclassifier', KNeighborsClassifier(n_neighbors=30))])

In [55]:
#predicciones
predictions = knn_pipe.predict(X_test)
predictions



array([2, 0, 0, ..., 0, 0, 2])

In [56]:
# Evaluacion en el conjunto de pruebas
# calculen la exactitud de la clasificación
acc_score = knn_pipe.score(X_test, y_test)
acc_score


0.5507662835249042

# Matriz de confusion

In [57]:
pred=knn.predict(X_test)
matriz = confusion_matrix(pred,y_test)
print(matriz)

[[280 146 230]
 [  0   0   0]
 [119 187  82]]


/usr/local/lib/python3.7/dist-packages/sklearn/base.py:444: UserWarning: X has feature names, but KNeighborsClassifier was fitted without feature names
  f"X has feature names, but {self.__class__.__name__} was fitted without"


**Evaluacion de la Matriz**

In [69]:
print('El Accuracy tiene un valor de : {:.2f}'.format(accuracy_score(y_test, pred)))
print('_____________________________________________________________________________________________')
print('La (Micropromedio) Precision tiene un valor de : {:.2f}'.format(precision_score(y_test, pred, average='micro')))
print('El (Micropromedio) Recall tiene un valor de: {:.2f}'.format(recall_score(y_test, pred, average='micro')))
print('El (Micropromedio) F1-score tiene un valor de: {:.2f}\n'.format(f1_score(y_test, pred, average='micro')))
print('_____________________________________________________________________________________________')
print('La (Macropromedio) Precision tiene un valor de : {:.2f}'.format(precision_score(y_test, pred, average='macro')))
print('El (Macropromedio) Recall tiene un valor de: {:.2f}'.format(recall_score(y_test, pred, average='macro')))
print('El (Macropromedio) F1-score tiene un valor de: {:.2f}\n'.format(f1_score(y_test, pred, average='macro')))
print('_____________________________________________________________________________________________')
print('La (Promedio ponderado) Precision tiene un valor de : {:.2f}'.format(precision_score(y_test, pred, average='weighted')))
print('El (Promedio ponderado) Recall tiene un valor de: {:.2f}'.format(recall_score(y_test, pred, average='weighted')))
print('El (Promedio ponderado) F1-score tiene un valor de: {:.2f}\n'.format(f1_score(y_test, pred, average='weighted')))
print('_____________________________________________________________________________________________')
print('Reporte de Clasificacion:')
print(classification_report(y_test, pred, target_names=['Class 1', 'Class 2', 'Class 3']))

El Accuracy tiene un valor de : 0.35
_____________________________________________________________________________________________
La (Micropromedio) Precision tiene un valor de : 0.35
El (Micropromedio) Recall tiene un valor de: 0.35
El (Micropromedio) F1-score tiene un valor de: 0.35

_____________________________________________________________________________________________
La (Macropromedio) Precision tiene un valor de : 0.21
El (Macropromedio) Recall tiene un valor de: 0.32
El (Macropromedio) F1-score tiene un valor de: 0.26

_____________________________________________________________________________________________
La (Promedio ponderado) Precision tiene un valor de : 0.23
El (Promedio ponderado) Recall tiene un valor de: 0.35
El (Promedio ponderado) F1-score tiene un valor de: 0.27

_____________________________________________________________________________________________
Reporte de Clasificacion
              precision    recall  f1-score   support

     Class 1       0.

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetr

**R1. Las clases se ven bien balanceadas existiendo en promedio 3 pts de diferencia entre clases**

**R4. La precision es del del 55% aprox, para los datos de test, considerando los valores de KNN**